### PR TITLE
APEXCORE-413 - Collision between Sink.getCount() and SweepableReservoir.getCount()

### PR DESCRIPTION
**Pull request by vrozov**
@tweise Please review

### CODE REVIEW

File: engine/src/main/java/com/datatorrent/stram/stream/OiOStream.java [hunk 1]
```diff
@@ -29,11 +29,12 @@
  *
  * @since 0.3.5
  */
-public class OiOStream implements Stream, SweepableReservoir
```
Comment (tweise) [inline engine/src/main/java/com/datatorrent/stram/stream/OiOStream.java:32]: Why the change to OiOStream?

File: engine/src/main/java/com/datatorrent/stram/stream/InlineStream.java [hunk 2]
```diff
@@ -80,14 +88,29 @@ public void teardown()
   public void put(Object tuple)
   {
     try {
-      super.put(tuple);
+      reservoir.put(tuple);
+      if (!(tuple instanceof Tuple)) {
```
Comment (tweise) [inline engine/src/main/java/com/datatorrent/stram/stream/InlineStream.java:92]: What's the cost off adding the extra instanceof?

**Comment by tweise** [comment]
A more general question: Is there any benefit of AbstractReservoir implementing BlockingQueue and hence carrying all the extra methods that we don't need but every concrete implementation must provide?

File: engine/src/main/java/com/datatorrent/stram/stream/OiOStream.java (hunk 1 above)
Comment (vrozov) [inline engine/src/main/java/com/datatorrent/stram/stream/OiOStream.java:32]: It is necessary to maintain two different counts - one for Sink and one for SweepableReservoir as they are reset separately. In case, OioStream implements both Stream and SweepableReservoir, it will not be possible to distinguish which count should be reset to 0, so either `Sink.getCount()` or `SweepableReservoir.getCount()` would return a wrong result.

File: engine/src/main/java/com/datatorrent/stram/stream/InlineStream.java (hunk 2 above)
Comment (vrozov) [inline engine/src/main/java/com/datatorrent/stram/stream/InlineStream.java:92]: Please see http://stackoverflow.com/questions/103564/the-performance-impact-of-using-instanceof-in-java.

File: engine/src/main/java/com/datatorrent/stram/stream/OiOStream.java (hunk 1 above)
Comment (tweise) [inline engine/src/main/java/com/datatorrent/stram/stream/OiOStream.java:32]: OiOStream has only one thread operating on it, why is there a need for separate counts or issue with reset?

File: engine/src/main/java/com/datatorrent/stram/stream/InlineStream.java (hunk 2 above)
Comment (tweise) [inline engine/src/main/java/com/datatorrent/stram/stream/InlineStream.java:92]: I saw that :-) But since we are benchmarking plain stream performance and spending time and effort enhancing the same, we should probably also quantify this. Have you measured before/after?

File: engine/src/main/java/com/datatorrent/stram/stream/OiOStream.java (hunk 1 above)
Comment (vrozov) [inline engine/src/main/java/com/datatorrent/stram/stream/OiOStream.java:32]: Let's say upstream operator puts n tuples into OiOStream, both `Sink.getCount(true)` and `SweepableReservoir.getCount(true)` should return n and reset count to 0. If there is only one count, only one call (whichever comes first) would return n, the second one would return 0.

**Comment by vrozov** [comment]
There is no benefit of AbstractReservoir implementing BlockingQueue other than reusing already defined interface currently. As CircularBuffer implements BlockingQueue and DefaultReservoir extended CircularBuffer, I made AbstractReservoir implement BlockingQueue interface as well. As AbstractReservoir implementations delegate BlockingQueue implementation to CircularBuffer, SpscArrayQueue or ArrayBlockingQueue, and both CircularBuffer and ArrayBlockingQueue implement BlockingQueue, if SpscArrayQueue adds support for BlockingQueue, it may be possible to use other methods from BlockingQueue in AbstractReservoirs.

File: engine/src/main/java/com/datatorrent/stram/stream/InlineStream.java (hunk 2 above)
Comment (vrozov) [inline engine/src/main/java/com/datatorrent/stram/stream/InlineStream.java:92]: I don't see any difference in InlineStream performance before and after the change.

File: engine/src/main/java/com/datatorrent/stram/stream/OiOStream.java (hunk 1 above)
Comment (tweise) [inline engine/src/main/java/com/datatorrent/stram/stream/OiOStream.java:32]: How about moving this as documentation into the javadoc block :-)

File: engine/src/main/java/com/datatorrent/stram/stream/InlineStream.java (hunk 2 above)
Comment (tweise) [inline engine/src/main/java/com/datatorrent/stram/stream/InlineStream.java:92]: Is that because of correct counting or because instanceof has no impact :-)
